### ISSUE TITLE
Temperature-aware YOLO pipeline using animal/body profiles to reduce color-map hallucinations

### ISSUE THREAD

**Issue by luckiday** [AIoT-Infrastructure/pythermal#5]
## Summary

Extend `pythermal` with a **temperature-aware object detection pipeline** that combines:

1. **Radiometric temperature data** (absolute temps from the sensor), and  
2. **Animal/body temperature profiles** (e.g., human, dog, cat),  

to improve YOLO-based recognition on thermal images and **reduce color-map hallucinations**, e.g.:

- Current issue: YOLO trained on pseudo-color images tends to:
  - Misrecognize **hot blobs** (e.g., heater, reflection) as human heads,
  - Overfit to **color** (e.g., “orange = head”) instead of actual thermal structure and plausible temperature ranges.

Goal: Provide reusable utilities + reference code so that downstream users can build **temperature-constrained detectors** instead of “RGB-only YOLO on pseudo-color thermal”.

---

## Background / Problem

Right now, typical YOLO workflows on thermal use:

- Pseudo-color thermal images (e.g., 240×240 RGB),
- Standard YOLO backbones (trained or fine-tuned on these colorized maps).

This has several problems:

1. **Color-map dependency**
   - If the color map changes (e.g., “iron” → “rainbow”), YOLO performance may collapse.
   - YOLO mistakes high-intensity color (e.g., orange, yellow) for semantic meaning (“this must be a head”).

2. **No use of absolute temperature**
   - Thermal sensor provides **radiometric data** (e.g., 16-bit absolute temps), but YOLO sees only 0–255 RGB after colormap.
   - Model cannot enforce: “human head should be in ~[32°C, 37°C] and have this kind of spatial pattern”.

3. **No physical plausibility check**
   - Outputs are not validated against any **body temperature profile**:
     - A “human” box over a region at 10°C is accepted as valid.
     - Heaters at 60°C are happily classified as “person” or “head”.

The new animal/body temperature profile database (planned or existing `pythermal.animals`) gives us a chance to **inject physics + physiology** into detection.

---

## High-level Goals

- Use temperature profiles to **guide YOLO** at:
  1. **Input level** (extra channels / masks),
  2. **Post-processing level** (temperature-aware filtering / re-scoring),
  3. **Training-time augmentation / loss weighting** (optional, later).

- Provide a minimal, well-documented **reference pipeline** that:
  - Can be used for human detection (eldercare, occupancy, safety),
  - Can be extended for pets / livestock later.

---

### 1. Multi-channel YOLO input (temp + pseudo-color)

Instead of feeding only RGB pseudo-color images:

**Input:**
- `C1`: pseudo-color thermal image (H×W×3, uint8)
- `C2`: normalized absolute temperature map (H×W×1, float32)
- optionally `C3`: temperature-likelihood mask based on `AnimalTempProfile`

Implementation approach:

- Preprocessing function:

  ```python
  from pythermal.yolo import build_temperature_aware_input

  image_tensor = build_temperature_aware_input(
      raw_temp_frame,     # H×W float32
      colormap="iron",    # or whoever we use
      species="human",    # to build optional temp-likelihood mask
      resize=(320, 320)   # YOLO input size
  )
  # -> torch.Tensor with shape (C, H, W)
  ````

* `build_temperature_aware_input`:

  * Generates pseudo-color RGB from `raw_temp_frame`.
  * Normalizes temperature to [0, 1] and concatenates as an extra channel.
  * Optionally builds a “probability mask” that highlights temps in the expected human body range and adds as another channel.

This can be prototyped as a module with clear boundaries; actual YOLO training can live in a separate repo/notebook.

---

### 2. Temperature-aware post-processing of YOLO detections

Given YOLO predictions:

```python
detections = yolo_model(image_tensor)
# list[ {bbox, class_id, score, ...}, ... ]
```

We add a post-processing step:

```python
from pythermal.yolo import temperature_aware_filter

filtered_detections = temperature_aware_filter(
    detections,
    raw_temp_frame,
    species="human",
    temp_profiles_db=...
)
```

`temperature_aware_filter` does:

1. For each bbox, extract the temperature patch from `raw_temp_frame`.

2. Compute statistics: mean, median, max, min.

3. Look up species profile:

   ```python
   profile = get_profile("human")
   core_range = profile.typical_core_range  # e.g., (36.5, 37.5)
   surface_offset = profile.typical_surface_offset  # e.g., (-4, 0)
   ```

4. Derive an **expected temp range** for surface/head, e.g.:

   ```python
   expected_min = core_min + surface_offset_min
   expected_max = core_max + surface_offset_max
   ```

5. Re-score or filter detections:

   * If bbox temps are far outside expected range, down-weight or discard.
   * If bbox temps are consistent with human head/torso, slightly boost the score.

This directly targets the problem: “YOLO recognizes hot, saturated orange region as a head”. If a heater at 55°C is detected as “person”, the temperature-aware filter will strongly penalize it.

---

### 3. (Optional, advanced) Temperature-aware loss weighting

For contributors who want to go deeper:

* Define a custom loss or regularizer that encourages:

  * Detected “human” regions to have temperature statistics within the plausible range.
  * Background-only regions (e.g., walls) not to receive strong “human” predictions.

This could be done by adding a **temperature consistency term**:

```python
L_temp = sum_over_boxes( f(temps_in_box, species_profile) )
```

and combining:

```python
L_total = L_yolo + λ * L_temp
```

This part can be scoped as a “research / experimental” subtask and may live in a `research/` directory or external notebook.

---

## Task Breakdown

This feature is advanced and can be split among several contributors:

### Phase 1: Infrastructure

* [ ] Implement `build_temperature_aware_input(raw_temp_frame, colormap, species, resize)`:

  * [ ] Generate pseudo-color image.
  * [ ] Normalize temp map and stack as extra channel.
  * [ ] (Optional) Generate temperature-likelihood mask from `AnimalTempProfile`.
  * [ ] Provide tests with synthetic temp maps.

* [ ] Design simple `pythermal.yolo` namespace (or similar) for these utilities.

### Phase 2: Temperature-aware post-processing

* [ ] Implement `temperature_aware_filter(detections, raw_temp_frame, species, ...)`.
* [ ] Plug into an example YOLO inference script:

  * [ ] Before/after comparison on sample videos or frame sequences.
  * [ ] Add logging/visualization of filtered-out boxes.

### Phase 3: Example + Documentation

* [ ] Create a notebook `examples/yolo_temperature_aware.ipynb` showing:

  * [ ] The current problem: YOLO misdetects hot objects as heads.
  * [ ] How temperature-aware filtering reduces false positives.
  * [ ] Visualization overlays: bbox + temperature statistics.

* [ ] Document the design in `docs/yolo_temperature_aware.md`:

  * [ ] Explain color-map hallucination problem.
  * [ ] Describe animal/body profile integration.
  * [ ] Show how to plug these utilities into a user’s existing YOLO pipeline.

### Phase 4 (Optional): Experimental loss

* [ ] Prototype a simple `temperature_consistency_loss(...)` in a training script.
* [ ] Run a small experiment on a toy dataset and log results (even if small-scale).

---

## Acceptance Criteria

* A `pythermal.yolo` (or similar) module exists with:

  * [ ] `build_temperature_aware_input(...)`
  * [ ] `temperature_aware_filter(...)`

* Temperature profiles from `pythermal.animals` are actually used in:

  * [ ] Post-processing to filter or re-score detections.
  * [ ] Optional temperature-likelihood mask for input.

* At least one worked example (script or notebook) demonstrates:

  * [ ] YOLO misclassifying “orange hot blobs” as humans.
  * [ ] Improved precision after temperature-aware filtering, with qualitative examples.

* Documentation clearly states:

  * [ ] Limitations (e.g., depends on sensor calibration, emissivity, etc.).
  * [ ] How users can disable / customize temperature constraints.
  * [ ] How to plug in new species profiles (e.g., pets, livestock).

---

## Notes

* This is meant to be a **bridge between physics and deep learning**:

  * Uses existing YOLO implementation,
  * Adds temperature-based priors derived from our profile database.
